libcontainer/container_linux.go

```diff
@@ -779,6 +779,9 @@ const descriptorsFilename = "descriptors.json"
 
 func (c *linuxContainer) addCriuDumpMount(req *criurpc.CriuReq, m *configs.Mount) {
 	mountDest := strings.TrimPrefix(m.Destination, c.config.Rootfs)
+	if dest, err := securejoin.SecureJoin(c.config.Rootfs, mountDest); err == nil {
+		mountDest = dest[len(c.config.Rootfs):]
```

Comment:
I do wonder -- what happens if the mount is moved from inside the container (with a simple `rename`) after the filesystem has been set up? Is this also something CRIU doesn't support / really doesn't like?

Reply:
Yes @cyphar  Since CRIU does not know how to restore bind mounts(master in host namespace),  CRIU need  this type of mount marked as external(via `--external`). 
  runc formats `--external` parameter according to spec((link)[https://github.com/opencontainers/runc/blob/master/libcontainer/container_linux.go#L1112]). So for situation you metioned,  criu does not work properly.  btw, should we export `--external` for runc. Then users can mark external resource themselves?

Reply:
>> I do wonder -- what happens if the mount is moved from inside the container (with a simple rename) after the filesystem has been set up? Is this also something CRIU doesn't support / really doesn't like?

I do not understand this question. But initially I would have said that CRIU can handle it. Now @liusdu says it is not possible. I probably misunderstood the question.

> So for situation you mentioned, CRIU does not work properly. btw, should we export `--external` for runc. Then users can mark external resource themselves?

For Kubernetes, restoring containers in different Pods, I will need CRIU's `--external` to be available at the runc level.

Currently runc informs CRIU about all external mounts but it is not possible to change the location of the external mount between checkpoint and restore. For Kubernetes I will need to change the location of an external mount. The most obvious bind mount currently is the location of the secrets directory which has the Pod UID in its container external path and I need to tell CRIU/runc that that directory should be mounted from some other location.

Reply:
ping @cyphar  ..